sdk/cosmos/azure_data_cosmos/src/query/executor.rs

```diff
@@ -0,0 +1,202 @@
+    phantom: std::marker::PhantomData<fn() -> T>,
+}
+
+impl<T: DeserializeOwned + 'static> QueryExecutor<T> {
+    pub fn new(
+        http_pipeline: CosmosPipeline,
+        container_link: ResourceLink,
+        query: Query,
+        options: QueryOptions<'_>,
+        query_engine: QueryEngineRef,
+    ) -> azure_core::Result<Self> {
+        let items_link = container_link.feed(ResourceType::Items);
+        let context = options.method_options.context.into_owned();
+        let base_request =
+            pipeline::create_base_query_request(http_pipeline.url(&items_link), &query)?;
+        Ok(Self {
+            http_pipeline,
+            container_link,
+            items_link,
+            context,
+            query_engine,
+            base_request,
+            query,
+            pipeline: None,
+            phantom: std::marker::PhantomData,
+        })
+    }
+
+    pub fn into_stream(self) -> azure_core::Result<FeedPager<T>> {
+        Ok(FeedPager::from_stream(futures::stream::try_unfold(
+            self,
+            |mut state| async move {
+                let val = state.step().await?;
+                Ok(val.map(|item| (item, state)))
+            },
+        )))
+    }
+
+    /// Executes a single step of the query execution.
+    ///
+    /// # Returns
+    ///
+    /// An item to yield, or None if execution is complete.
+    #[tracing::instrument(skip_all)]
+    async fn step(&mut self) -> azure_core::Result<Option<FeedPage<T>>> {
+        let pipeline = match self.pipeline.as_mut() {
+            Some(pipeline) => pipeline,
+            None => {
+                // Initialize the pipeline.
+                let query_plan = get_query_plan(
+                    &self.http_pipeline,
+                    &self.items_link,
+                    self.context.to_borrowed(),
+                    &self.query,
+                    self.query_engine.supported_features()?,
+                )
+                .await?
+                .into_raw_body()
+                .collect()
+                .await?;
+                let pkranges = get_pkranges(
+                    &self.http_pipeline,
+                    &self.container_link,
+                    self.context.to_borrowed(),
+                )
+                .await?
+                .into_raw_body()
+                .collect()
+                .await?;
+
+                let pipeline =
+                    self.query_engine
+                        .create_pipeline(&self.query.text, &query_plan, &pkranges)?;
+                self.pipeline = Some(pipeline);
+                self.pipeline.as_mut().expect("we just set it")
+            }
+        };
+
+        // Execute the pipeline to get a page of results
+        while !pipeline.complete() {
+            // Run the pipeline
+            let results = pipeline.run()?;
```

Comment:
Since `complete()` returns immediately, are we basically just taking advantage of any async waits in `pipeline.run()`? Otherwise, this seems very CPU-intensive and wasteful.